JotunnLib/Managers/DungeonManager.cs

```diff
@@ -0,0 +1,387 @@
+        /// </summary>
+        /// <param name="prefab">The <see cref="DungeonGenerator"/> prefab.</param>
+        /// <param name="themeName">The name of the theme to register.</param>
+        /// <returns>True if theme is successfullly registered, otherwise false.</returns>
+        public bool RegisterDungeonTheme(GameObject prefab, string themeName)
+        {
+            Logger.LogDebug($"RegisterDungeonTheme called with prefab {prefab.name} and themeName {themeName}.");
+
+            if (prefab == null)
+            {
+                throw new ArgumentException("Cannot be null", nameof(prefab));
+            }
+            if (string.IsNullOrEmpty(themeName))
+            {
+                throw new ArgumentException("Cannot be null or empty", nameof(themeName));
+            }
+
+            DungeonGenerator dg = prefab.GetComponentInChildren<DungeonGenerator>();
+            if (dg == null)
+            {
+                Logger.LogError($"Cannot find DungeonGenerator component in prefab {prefab.name}.");
+                throw new ArgumentException("Prefab must contain a DungeonGenerator component", "prefab");
+            }
+
+            DungeonGeneratorTheme dgt;
+            if (dg.gameObject.TryGetComponent(out dgt))
+            {
+                if (!string.IsNullOrEmpty(dgt.m_themeName) && dgt.m_themeName != themeName)
+                {
+                    Logger.LogWarning($"Overwriting existing theme name {dgt.m_themeName} with {themeName}.");
+                }
+            }
+            else
+            {
+                dgt = dg.gameObject.AddComponent<DungeonGeneratorTheme>();
+            }
+            dgt.m_themeName = themeName;
+            themeList.Add(themeName);
+
+            return true;
+        }
+
+
+        /// <summary>
+        ///     Adds a new environment prefab to be registered when <see cref="ZoneSystem.SetupLocations"/> runs.<br />
+        ///     If you intend to use a custom interior environment <see cref="Location.m_interiorEnvironment"/>, this method enables you 
+        ///     to provide a prefab with an appropriately configured <see cref="LocationList"/> containing atleast one
+        ///     <see cref="EnvSetup"/> within <see cref="LocationList.m_environments"/>.
+        /// </summary>
+        /// <param name="assetBundle">The <see cref="AssetBundle"/> containing the prefab.</param>
+        /// <param name="prefabName">The name of the prefab to register.</param>
+        public void RegisterEnvironment(AssetBundle assetBundle, string prefabName)
+        {
+            if (assetBundle == null)
+            {
+                throw new ArgumentException("Cannot be null", nameof(assetBundle));
+            }
+            if (string.IsNullOrEmpty(prefabName))
+            {
+                throw new ArgumentException("Cannot be null or empty", nameof(prefabName));
+            }
+            envNames.Add(prefabName, assetBundle);
```

Comment:
prefabs should be loaded early when possible, mostly to give direct feedback if a prefab is missing or an AssetBundle is invalid.

Reply:
@MSchmoecker Could you further clarify what needs to change here? I'm not sure what I should change in this code from your comment.

Reply:
The AssetBundle is loaded at later time than the `RegisterEnvironment` call: https://github.com/Valheim-Modding/Jotunn/blob/24ce92e2ccb448652ea7e449cded4fd1ccc923f8/JotunnLib/Managers/DungeonManager.cs#L257-L267

The loading part should be moved to RegisterEnvironment, while the `env.FixReferences(true)` should stay at OnZoneSystemSetupLocations to make sure vanilla assets are loaded.

Looking at this code part, `UnityEngine.Object.Instantiate<GameObject>(env)` should also changed to spawn under a `DontDestroyOnLoad` object to not be destroyed on scene change. See https://github.com/Valheim-Modding/Jotunn/blob/294c8304f3d6ff1270420eb64a395be98aa8b16c/JotunnLib/Managers/PrefabManager.cs#L69-L71 and related code

Reply:
Addressed in: https://github.com/Valheim-Modding/Jotunn/pull/434/commits/9338dc963c4127a7e81e3f7f8bdfc267c6ffe323